File: server/src/api/sources.rs
//! `/api/sources` — port of `Api/Sources.cs` (ADR 0033, S3 #125). List is
//! auth-only; add/patch/delete are admin-gated. `POST` stays a 502 stub
//! until the plugin system is ported (ADR 0030) — matches .NET exactly.

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::routing::get;
use axum::{Json, Router};
use serde::{Deserialize, Serialize};

use crate::auth::Claims;
use crate::error::{AppError, AppResult};
use crate::sources;
use crate::state::AppState;

pub fn routes() -> Router<AppState> {
    Router::new().route("/", get(list).post(add)).route(
        "/{id}",
        axum::routing::patch(patch_source).delete(delete_source),
    )
}

#[derive(Serialize)]
struct SourceRowDto {
    id: String,
    name: String,
    base_url: String,
    has_api_key: bool,
    content_types: Vec<String>,
    enabled: bool,
    is_community: bool,
    priority: i64,
}

impl From<sources::SourceRow> for SourceRowDto {
    fn from(s: sources::SourceRow) -> Self {
        Self {
            id: s.id,
            name: s.name,
            base_url: s.base_url,
            has_api_key: s.api_key.is_some(),
            content_types: s
                .content_types
                .split(',')
                .map(str::trim)
                .filter(|s| !s.is_empty())
                .map(String::from)
                .collect(),
            enabled: s.enabled,
            is_community: s.is_community,
            priority: s.priority,
        }
    }
}

async fn list(
    _claims: Claims,
    State(state): State<AppState>,
) -> AppResult<Json<Vec<SourceRowDto>>> {
    let rows = sources::list_ordered_by_created_at(&state.db).await?;
    Ok(Json(rows.into_iter().map(SourceRowDto::from).collect()))
}

#[derive(Deserialize)]
struct AddSourceBody {
    #[allow(dead_code)]
    base_url: String,
    #[allow(dead_code)]
    api_key: Option<String>,
}

/// Stub — must probe the plugin host to get name/content_types/default_explicit.
/// Wired when the plugin system is ported (ADR 0030).
async fn add(claims: Claims, Json(_body): Json<AddSourceBody>) -> AppResult<StatusCode> {
    claims.require_admin()?;
    Ok(StatusCode::BAD_GATEWAY)
}

#[derive(Deserialize)]
struct PatchSourceBody {
    #[serde(default)]
    enabled: bool,
    priority: Option<i64>,
}

async fn patch_source(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(body): Json<PatchSourceBody>,
) -> AppResult<StatusCode> {
    claims.require_admin()?;
    sources::find_by_id(&state.db, &id)
        .await?
        .ok_or(AppError::NotFound)?;

    sources::update_enabled(&state.db, &id, body.enabled).await?;
    if let Some(priority) = body.priority {
        sources::update_priority(&state.db, &id, priority).await?;
    }

    Ok(StatusCode::NO_CONTENT)
}

async fn delete_source(
    claims: Claims,
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> AppResult<StatusCode> {
    claims.require_admin()?;
    if !sources::delete(&state.db, &id).await? {
        return Err(AppError::NotFound);
    }
    Ok(StatusCode::NO_CONTENT)
}

File: server/Api/Sources.cs
using System.Security.Claims;
using System.Text.Json.Serialization;
using ArrghServer.Data;
using ArrghServer.Data.Models;
using Microsoft.EntityFrameworkCore;

namespace ArrghServer.Api;

public static class Sources
{
    public static RouteGroupBuilder MapSourcesRoutes(this RouteGroupBuilder group)
    {
        group.MapGet("/", ListSources).RequireAuthorization().WithSummary("List configured plugin sources");
        group.MapPost("/", AddSource).RequireAuthorization().WithSummary("Add a plugin source");
        group.MapPatch("/{id}", PatchSource).RequireAuthorization().WithSummary("Update source (enabled, priority, api_key…)");
        group.MapDelete("/{id}", DeleteSource).RequireAuthorization().WithSummary("Remove a source");
        return group;
    }

    // -------------------------------------------------------------------------

    static async Task<IResult> ListSources(ClaimsPrincipal principal, AppDbContext db)
    {
        // Two-step: EF loads raw fields, then client-side split of content_types CSV
        var raw = await db.ExternalSources
            .OrderBy(s => s.CreatedAt)
            .Select(s => new { s.Id, s.Name, s.BaseUrl, s.ApiKey, s.ContentTypes, s.Enabled, s.IsCommunity, s.Priority })
            .ToListAsync();

        var rows = raw.Select(s => new SourceRowDto
        {
            Id = s.Id,
            Name = s.Name,
            BaseUrl = s.BaseUrl,
            HasApiKey = s.ApiKey is not null,
            ContentTypes = s.ContentTypes
                .Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries),
            Enabled = s.Enabled,
            IsCommunity = s.IsCommunity,
            Priority = s.Priority,
        }).ToList();

        return Results.Ok(rows);
    }

    static Task<IResult> AddSource(AddSourceBody body, ClaimsPrincipal principal, IConfiguration config)
    {
        if (!IsAdmin(principal)) return Task.FromResult(Results.Forbid());

        // RAW HTTP — intentional: must probe plugin host to get source name/content_types/default_explicit.
        // Stubbed until plugin system is ported (ADR 0030).
        // When ported: GET {PluginHostUrl}/plugins/{sourceKey}/info → populate ExternalSource fields.
        return Task.FromResult(Results.StatusCode(502));
    }

    static async Task<IResult> PatchSource(string id, PatchSourceBody body, ClaimsPrincipal principal, AppDbContext db)
    {
        if (!IsAdmin(principal)) return Results.Forbid();

        var source = await db.ExternalSources.FindAsync(id);
        if (source is null) return Results.NotFound();

        source.Enabled = body.Enabled;
        if (body.Priority.HasValue) source.Priority = body.Priority.Value;
        await db.SaveChangesAsync();
        await ReloadRegistryAsync();

        return Results.NoContent();
    }

    static async Task<IResult> DeleteSource(string id, ClaimsPrincipal principal, AppDbContext db)
    {
        if (!IsAdmin(principal)) return Results.Forbid();

        var source = await db.ExternalSources.FindAsync(id);
        if (source is null) return Results.NotFound();

        db.ExternalSources.Remove(source);
        await db.SaveChangesAsync();
        await ReloadRegistryAsync();

        return Results.NoContent();
    }

    // -------------------------------------------------------------------------

    internal static bool IsAdmin(ClaimsPrincipal p) =>
        p.FindFirstValue(ClaimTypes.Role) == "admin";

    // TODO: rebuild in-memory source registry when plugin system is ported (ADR 0030)
    static Task ReloadRegistryAsync() => Task.CompletedTask;
}

// ── DTOs ──────────────────────────────────────────────────────────────────────

public class SourceRowDto
{
    public string Id { get; set; } = null!;
    public string Name { get; set; } = null!;
    public string BaseUrl { get; set; } = null!;
    public bool HasApiKey { get; set; }
    public string[] ContentTypes { get; set; } = [];
    public bool Enabled { get; set; }
    public bool IsCommunity { get; set; }
    public int Priority { get; set; }
}

public class AddSourceBody
{
    [JsonPropertyName("base_url")] public string BaseUrl { get; set; } = null!;
    [JsonPropertyName("api_key")]  public string? ApiKey { get; set; }
}

public class PatchSourceBody
{
    [JsonPropertyName("enabled")]  public bool Enabled { get; set; }
    [JsonPropertyName("priority")] public int? Priority { get; set; }
}
